File: BinaryTrees/max_depth.py
from typing import Optional

from tree_node import TreeNode


class MaxDepth:
    def max_depth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        return 1 + max(self.max_depth(root.left), self.max_depth(root.right))


    # Time complexity: O(n) where n is the number of nodes in the binary tree. We visit each node once.
    # Space complexity: O(h) where h is the height of the binary tree. 
    # In the worst case of a completely unbalanced tree, the height could be n, leading to O(n) space complexity. 
    # In the best case of a balanced tree, the height would be log(n), leading to O(log(n)) space complexity.

    def max_depth_iterative(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        stack = [(root, 1)]
        max_depth = 0
        while stack:
            node, depth = stack.pop()
            max_depth = max(max_depth, depth)
            if node.left:
                stack.append((node.left, depth + 1))
            if node.right:
                stack.append((node.right, depth + 1))

        return max_depth

    # Time complexity: O(n) where n is the number of nodes in the binary tree. We visit each node once.
    # Space complexity: O(h) where h is the height of the binary tree. 
    # In the worst case of a completely unbalanced tree, the height could be n, leading to O(n) space complexity. 
    # In the best case of a balanced tree, the height would be log(n), leading to O(log(n)) space complexity.

File: BinaryTrees/maxDepth.js
/*Given the root of a binary tree, return its maximum depth.

A binary tree's maximum depth is the number of nodes along the longest path from the root node down to the farthest leaf node.*/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

const maxDepthRecursive = (root) => {
  if (!root) {
    return 0;
  }

  const left = maxDepth(root.left);
  const right = maxDepth(root.right);

  return Math.max(left, right) + 1;
};

const maxDepthIterative = (root) => {
  if (!root) {
    return 0;
  }

  let maxDepth = 0;

  const stack = [];
  stack.push([root, 1]);

  while (stack.length) {
    let [currentNode, currentDepth] = stack.pop();

    maxDepth = Math.max(maxDepth, currentDepth);

    if (currentNode.left) stack.push([currentNode.left, currentDepth + 1]);
    if (currentNode.right) stack.push([currentNode.right, currentDepth + 1]);
  }

  return maxDepth;
};

// Time: O(n) where n is the number of nodes in the tree
// Space: O(n) in the worst case of a skewed tree, O(log n) in the best case of a balanced tree